3D Tiles refactoring

jailln opened the pull request:
## Description
Some refactoring of 3D Tiles management to bring it closer to the specification. Main changes are (see commits for a detailed list of changes):

* Update naming convention to match the one that seemed to have been chosen for the 3D Tiles layer (prefixing with C3DTiles, that I simplified to C3DT for core classes).
* Start creating classes matching the specification (C3DTileset, C3DTBatchTable, C3DTBoundingVolume)
* Various attributes renaming (e.g. tileIndex -> tileset, index -> tiles)
* Update 3D Tiles extensions management: they are now registered in the layer with their name and an object holding a parsing method and other methods such as `getPickingInfo`. 
* Update 3D Tiles related tests and documentation.

Some of the doc doesn't seem to be generated (for classes: C3DTileset, C3DTBatchTable, C3DTBoundingVolume and C3DTBatchTableHierarchyExtension) while it does for C3DTExtensions and C3DTAbstractExtension; any ideas why?

## Motivation and Context

This PR starts fixing issue #1171

jailln committed:
refactor(3dtiles):Add spec classes and modify extensions management.

* BREAKING CHANGE: 3D Tiles Tileset index attribute (stored in
layer.tileIndex.index) is now an array starting at 0.

* BREAKING CHANGE: 3D Tiles extensions are now registered in the
registeredExtensions attribute of C3DTilesLayer. See
examples/3dtiles_batch_table.html for an example.

* BREAKING CHANGE: C3DTExtensions expects for an object extending
C3DTAbstractExtension to manage a registered extension instead
of a parser.

* BREAKING CHANGE: Rename core 3D Tiles classes: $3DTilesExtensions
becomes C3DTExtensions and $3dTilesAbstractExtension becomes
C3DTAbstractExtension and BatchTableHierarchyExtension becomes
C3DTBatchTableHierarchyExtension.

* New classes maching the 3D Tiles specification:
C3DTBatchTable, C3DTBoundingVolume, C3DTileset.

* Add attributes to 3D Tiles core classes indicating their type

* 3D Tiles tests updated.

* 3D Tiles documentation updated.

jailln committed:
refactor(3dtiles): rename tileIndex property of C3DTilesLayer to tileset.

BREAKING CHANGE: C3DTilesLayer tileIndex renamed to tileset.

BREAKING CHANGE: C3DTileset's index attribute renamed to tiles.

jailln committed:
refactor(3d tiles): move attribute from layer to tileset to match spec.

jailln committed:
chore(3dtiles): check that extensions found in a tileset have been correctly registered to layer

gchoqueux commented:
Thank you for this contribution it looks very promising!

#### Before we start the review, I have a few checks and precisions.

* prefixing class with `C3DTile`  ✔️ 
* start creating classes matching the specification ✔️ 
    * To facilitate the reviewing, could you specify if there are **algorithms rewrites** or **algorithms moves** to their corresponding class?
* Various attributes renaming ✔️ 
* Update 3D Tiles extensions management: they are now registered in the layer with their name and an object holding a parsing method and other methods such as `getPickingInfo`.
   * I need to take a closer look, but generally `picking `is usually decorrelated from the data 
* Update 3D Tiles related tests and documentation. ✔️ 


#### Could you squash some commits to simplify the history ?
and move all messages commit in squashed commit.
I can participate in this work
by example:
  - chore(3dtiles): add check on bounding volume attribute.
  - split this commit : chore(3dtiles): clean C3DTilesSource and C3DTilesLayer.
  - chore(3dtiles): clean and document C3DTileset class.

#### Generate documentation

> Some of the doc doesn't seem to be generated (for classes: C3DTileset, C3DTBatchTable, C3DTBoundingVolume and C3DTBatchTableHierarchyExtension) while it does for C3DTExtensions and C3DTAbstractExtension; any ideas why?

@zarov , have you any ideas?

zarov commented:
> Some of the doc doesn't seem to be generated (for classes: C3DTileset, C3DTBatchTable, C3DTBoundingVolume and C3DTBatchTableHierarchyExtension) while it does for C3DTExtensions and C3DTAbstractExtension; any ideas why?

You need to declare your classes like this:
```js
class C3DTBatchTable {
    // ...
}

export default C3DTBatchTable;
```

I'm not sure why your version does not work, but this way all classes are the same.

jailln commented:
@gchoqueux :

**Clarifications:**

> * start creating classes matching the specification ✔️
>   * To facilitate the reviewing, could you specify if there are algorithms rewrites or algorithms moves to their corresponding class?

It is mainly code moving and sometimes simplification and renaming but without changes to the core algorithm.

> * Update 3D Tiles extensions management: they are now registered in the layer with their name and an object holding a parsing method and other methods such as getPickingInfo.
>     * I need to take a closer look, but generally picking is usually decorrelated from the data

The idea behind this choice is to allow one to register an object taking care of everything that is related with this extension, including parsing or getting the information of relative to this extension that can be displayed to a user; so that the code relative to an extension is autonomous and can be "plugged" to itowns from an external project.

**Commits organization:**

> Could you squash some commits to simplify the history ?
> and move all messages commit in squashed commit.
> I can participate in this work
> by example:
> 
>     chore(3dtiles): add check on bounding volume attribute.
>     split this commit : chore(3dtiles): clean C3DTilesSource and C3DTilesLayer.
>     chore(3dtiles): clean and document C3DTileset class.

Ok for the commit to split. Regarding the other ones: should I squash them with their respective previous commits ?

**Documentation:**

@zarov, thanks it worked :+1:

gchoqueux commented:
> It is mainly code moving and sometimes simplification and renaming but without changes to the core algorithm.

Very good

> The idea behind this choice is to allow one to register an object taking care of everything that is related with this extension, including parsing or getting the information of relative to this extension that can be displayed to a user; so that the code relative to an extension is autonomous and can be "plugged" to itowns from an external project.

ok

> Ok for the commit to split. Regarding the other ones: should I squash them with their respective previous commits ?

If possible, the idea is to minimize the number of commits for one file and describe all changes in the commit message.

jailln commented:
@gchoqueux done; let me know if more squashing are needed.

gchoqueux commented:
Yeah, thanks, that's fine. Just squashes those last few commits.
Because they are 4 commits with minor modifications (`console`, missing properties, docs...)

* Keep **refactor(3dtiles):Add spec classes and modify extensions management.**  squashes with :
   * chore(3dtiles): clean C3DTilesSource
   * chore(3dtiles): Remove unnecessary and temporary C3DTilesLayer attribute
   * chore(3dtiles): add warning for unsupported bounding volumes types.
   * refactor(3dtiles): Add property to 3D Tiles core classes indicating their type

jailln commented:
Done.

gchoqueux commented:
Ok, I'm gonna start reviewing

gchoqueux commented on src/Core/3DTiles/C3DTBatchTable.js:
```diff
@@ -0,0 +1,74 @@
+        // 3dtiles_hierarchy.html)
+        if (json.extensions) {
+            this.extensions =
+                registeredExtensions.parseExtensions(json.extensions, this);
```
If extensions were filtered, there would be no need to pass context. 
We could type the extensions.

By example in `C3DTBatchTableHierarchyExtension` and `C3DTBatchTable` :

Add property `C3DTBatchTableHierarchyExtension.type = 'batchTableExtension'`
then in `BatchTableParser` we could filter `registeredExtensions` by this type,
before passing it in to the `C3DTBatchTable` constructor.

We could also merge the `BatchTableParser.parse` and `C3DTBatchTable.constructor` because
this parser instances `C3DTBatchTable`.

jailln replied:
I am not sure to understand the beginning of your comment regarding adding a `type` attribute to the Extensions classes (e.g. `C3DTBatchTableHierarchyExtension`) to avoid passing the context; could you elaborate a bit please?

I agree that we should remove the `BatchTableParser` and move its content to `C3DTBatchTable`; I initially planned to do it in another PR but I will do it here.

gchoqueux replied:
[see comment](https://github.com/iTowns/itowns/pull/1387#discussion_r427881753)

jailln replied:
`BatchTableParser` removed.

gchoqueux commented on src/Core/3DTiles/C3DTBatchTable.js:
```diff
@@ -0,0 +1,74 @@
+     * table. Object is formatted as follow:
+     * {BatchTable: {BatchTableFeatureProperties} }.
+     */
+    getPickingInfo(batchID) {
```
could you rename to `getInfoById` ?

jailln replied:
Ok

gchoqueux commented on src/Core/3DTiles/C3DTBatchTableHierarchyExtension.js:
```diff
@@ -0,0 +1,149 @@
+     * @returns {Object} - displayable information relative to this extension
+     * and for the feature with id=featureId and for its parents
+     */
+    getPickingInfo(featureId) {
```
could you rename `getInfoById` ?

jailln replied:
Ok

gchoqueux commented on src/Core/3DTiles/C3DTBatchTableHierarchyExtension.js:
```diff
@@ -0,0 +1,149 @@
+ * the instance in this class. Goal: Ease the retrieval of the properties of an
+ * instance.
+ */
+class C3DT_BTHExtension_BatchTable {
```
In class name `C3DT_BTHExtension_BatchTable`, the 2 letters **`BT`** also stands for **B**atch **T**able?

jailln replied:
Yes and the **H** for **H**ierarchy, as a shorthand for the name of the extension (https://github.com/CesiumGS/3d-tiles/tree/master/extensions/3DTILES_batch_table_hierarchy)

gchoqueux replied:
What bothers me is that it's a two-time `batch table` in the name.

jailln replied:
Yes, in my opinion the choice of the extension name is confusing but it is what they have chosen in the standard specification.

gchoqueux commented on src/Core/3DTiles/C3DTBatchTableHierarchyExtension.js:
```diff
@@ -0,0 +1,149 @@
+ * 3D Tiles
+ * [Batch Table Hierarchy Extension](https://github.com/AnalyticalGraphicsInc/3d-tiles/tree/master/extensions/3DTILES_batch_table_hierarchy)
+ */
+class C3DTBatchTableHierarchyExtension extends C3DTAbstractExtension {
```
`C3DTBatchTableHierarchyExtension` isn't a really extension?
because is parse/instance `C3DT_BTHExtension_BatchTable` and handle picking Info?

jailln replied:
The idea behind this structure is that one object is in charge of the extension (e.g. `C3DTBatchTableHierarchyExtension` is in charge of the 3DTILES_batch_table_hierarchy extension). Let's name it the extension manager. However, an extension can have data scattered in multiple classes of 3D Tiles (in this case it is only in the batch table but other extensions also have information in tileset, bounding volume, etc.). 
The idea is to have one class per 'piece' of extension (e.g. one class for the piece in the batch table, one class for the piece in the tileset) that is responsible for parsing and possibly doing other thing with this 'piece' of extension. The instance of this class corresponds to the parsed 'piece of extension' and is stored in the corresponding 3D Tiles class and instantiated by the extension manager depending on the calling context (i.e. the 3D Tiles class).

You can see an example implementation [here](https://github.com/jailln/UDV/tree/UDV-temporal/UDV-Core/src/Modules/Temporal/3DTILES_temporal) where we implement an extension of 3D Tiles (not yet in the standard spec) having data in multiple 3D Tiles classes (the entry point is [this file](https://github.com/jailln/UDV/blob/UDV-temporal/UDV-Core/src/Modules/Temporal/3DTILES_temporal/3DTemporalExtension.js)). 

Let me know if my explanation is not clear, I can work on a schema / a sequence diagram.

gchoqueux replied:
**Common response with this other [comment](https://github.com/iTowns/itowns/pull/1387#discussion_r427359755)**

`C3DTBatchTableHierarchyExtension` is in charge of `C3DT_BTHExtension_BatchTable` ?
Yes, I'd like a diagram to make it clear.

May be, it's possible to simplify, by example :
```js
// Extension class to register in C3DTExtensions
class fooExtension {
    constructor(json) {
        // Parsing json
    }
    // type extension to filter in context
    static get type() { return 'foo' };
}

// in  C3DTExtensions class
C3DTExtensions.parseExtensions(extensionsJSON, type) {
    const parsedExtensions = {};
    for (const extName in extensionsJSON) {
        if (this.isExtensionRegistered(extName)) {
            const extension = this.getExtension(extName);
            if (extension.type == type) {
                parsedExtensions[extName] = new extension(extensionsJSON[extName]);
            }
        }
    }
    return parsedExtensions;
}
```
wdym?

jailln replied:
In case of the `3DTILES_batch_table_hierarchy` extension, it may work but for extensions in general I think it won't. 3D Tiles extensions can indeed be divided in several parts; each part being stored in a 3D Tiles class. For instance, the [draco point compression extension](https://github.com/CesiumGS/3d-tiles/tree/master/extensions/3DTILES_draco_point_compression) is composed of two parts: one in the `featureTable` class of 3D Tiles and one in the `batchTable` class (see [its JSON schema](https://github.com/CesiumGS/3d-tiles/tree/master/extensions/3DTILES_draco_point_compression/schema)). The implementation therefore follows this idea: the extension is implemented by one class and each extension class is also implemented by one class. When a 3D Tiles class is parsed (e.g. batch table), if an extension part is found it is passed to the corresponding extension which calls the extension part class for the parsing. 

In this PR, `C3DTBatchTableHierarchyExtension` corresponds to the extension and `C3DT_BTHExtension_BatchTable` corresponds to the extension part stored in the batch table.

gchoqueux commented on src/Core/3DTiles/C3DTBoundingVolume.js:
```diff
@@ -0,0 +1,127 @@
+                sphere[2]), sphere[3]);
+    }
+
+    static boundingVolumeCulling(layer, boundingVolume, camera, tileMatrixWorld) {
```
If `boundingVolume` is instance of  `C3DTBoundingVolume`, the static function isn't necessary?

same comment for `viewerRequestVolumeCulling` function.

jailln replied:
Yep you're right; I removed the `static`

gchoqueux commented on src/Core/3DTiles/C3DTExtensions.js:
```diff
@@ -0,0 +1,118 @@
+    /**
+     * This class must not be instanciated directly.
+     */
+    constructor() {
```
`C3DTAbstractExtension` is more parser and picking handler?

This abstraction may not be necessary if it has no common properties and implemented functions?

`C3DTExtensions` could check the integrity `Extension` in `registerExtension`.
by example : handle  the methods  `parse` or `getPickingInfo` if doesn't exist.

jailln replied:
The idea behind `C3DTAbstractExtension` was to expose a clearer interface for extensions to users (and the associated documentation). But if you prefer I can remove the `C3DTAbstractExtension` and check the integrity of registered extensions in the `registerExtension` method.

gchoqueux replied:
Yes, I prefer a explanations in the documentation, unless it is possible to add to this class properties/methods common to all `Extensions`.

jailln replied:
Actually, I think the parse method may be shared by all extension; I'll push a proposition.

gchoqueux commented on src/Core/3DTiles/C3DTileset.js:
```diff
@@ -0,0 +1,112 @@
+     * in the C3DTilesLayer (see {@link C3DTExtensions}
+     */
+    parseTiles(tile, baseURL, parent, registeredExtensions) {
+        const inverseTileTransform = new THREE.Matrix4();
```
`inverseTileTransform ` could be instanced before  `class C3DTileset`

jailln replied:
You mean before the class declaration ? What would be the benefit ?

gchoqueux replied:
yes, before the class declaration because `inverseTileTransform` is instanced each call function and increases the memory so decreases performance to clean memory with the Garbage Collector.

jailln replied:
Ok, I'll move it. Just for my personal curiosity, do you know if this gain is quantifiable or if it has been quantified on simple cases?

gchoqueux replied:
With tools performances in Chrome you can track the garbage collector time (Major GC).
Instance Object only used in scope function increase major GC time.

jailln replied:
Ok, thanks.

gchoqueux commented on src/Core/3DTiles/C3DTileset.js:
```diff
@@ -0,0 +1,112 @@
+
+        // The only reason to store _worldFromLocalTransform is because of
+        // extendTileset where we need the transform chain for one tile.
+        tile._worldFromLocalTransform = tile.transform;
```
`tile._worldFromLocalTransform = tile.transform;` could be moved in `else` because is crushed if there is `parent._worldFromLocalTransform`

jailln replied:
Ok

gchoqueux commented on src/Core/3DTiles/C3DTileset.js:
```diff
@@ -0,0 +1,112 @@
+        tile.baseURL = baseURL;
+        if (tile.children) {
+            for (const child of tile.children) {
+                this.parseTiles(child, baseURL, tile, registeredExtensions);
```
I notice that the tree structure is flattened and that this is also the case in the master branch.

here:
```js
this.parseTiles(child, baseURL, tile, registeredExtensions);
```

All `child` and `tile`  will finally be added in `this.tiles`, is that normal?

jailln replied:
Yes it is already the case on the master branch. In addition to that, there is a tree structure with the nodes representing the tiles contents stored as threejs objects in the `object3d` attribute of the layer. (it is in fact a bit more complicated than this, see issue #1039 )

I don't know why this choice has been made in the first place but I think we may be able to simplify this and keep only one tree structure with the tiles and their content loaded as threejs objects (as discussed in the third item of issue #1171 ). I haven't looked into it yet though and I think it might be better to do it in a future PR ?

gchoqueux replied:
Yes, indeed a major refactor is necessary.

gchoqueux commented on src/Layer/C3DTilesLayer.js:
```diff
@@ -210,15 +42,31 @@ class C3DTilesLayer extends GeometryLayer {
         // custom cesium shaders are not functional;
         this.overrideMaterials = config.overrideMaterials !== undefined ? config.overrideMaterials : true;
         this.name = config.name;
+        this.registeredExtensions = config.registeredExtensions !== undefined ? config.registeredExtensions : new C3DTExtensions();
```
you could write:
```js
this.registeredExtensions = config.registeredExtensions || new C3DTExtensions();
```

jailln replied:
Ok

gchoqueux commented on src/Parser/BatchTableParser.js:
```diff
@@ -1,116 +1,22 @@
-        }
-    }
-}
+import C3DTBatchTable from '../Core/3DTiles/C3DTBatchTable';
```
you could write 
```js
import C3DTBatchTable from 'Core/3DTiles/C3DTBatchTable';
```

jailln replied:
Ok

gchoqueux reviewed (commented):
Thanks, this PR better structures the 3d tiles codes.

I'm doing a first global review that focuses on the structure of the classes that manage the `Extensions`.

jailln committed:
refactor(3dtiles): remove BatchTableParser
feat(3dtiles): parse pnts feature table

jailln commented:
I updated the extensions management as we discussed; let me know if it's ok (except from some documentation that I still need to update). I also took this opportunity to simplify `3DTilesHelper` and to move some code to `C3DTilesLayer`.

gchoqueux commented on src/Core/3DTiles/C3DTBatchTableHierarchyExtension.js:
```diff
@@ -69,46 +73,36 @@ class BatchTableHierarchyExtension extends $3dTilesAbstractExtension {
-        return Promise.resolve(new BatchTableHierarchyExtension(json));
-    },
-};
+function C3DT_BTHExtension_BatchTable_parser(json) {
```
This function shouldn't be necessary , could you directly register the class instead of `C3DT_BTHExtension_BatchTable_parser`?

jailln replied:
I can register an instance of the class and move the content of the constructor to a parse method of the class; would you prefer that ?

gchoqueux replied:
I meant : could register `C3DT_BTHExtension_BatchTable` instead of parser:

```js
extensions.registerExtension('3DTILES_batch_table_hierarchy',
        { batchtable: C3DT_BTHExtension_BatchTable });
```

jailln replied:
No this is not possible, I can only register a class instance or a function. Maybe registering a class instance is clearer than registering `C3DT_BTHExtension_BatchTable_parser`.

gchoqueux replied:
I've tried this and it works.

The registering:
```js
extensions.registerExtension("3DTILES_batch_table_hierarchy",
    { [itowns.C3DTilesTypes.batchtable]:  itowns.C3DT_BTHExtension_BatchTable });
```

The instancing in `C3DTExtensions#parseExtensions`
```js
parsedExtensions[extName] = new (this.getExtension(extName, type))(extensionsJSON[extName]);
```

jailln replied:
Nice, thanks ! I must have made a mistake when I tried it. It is simpler this way.

gchoqueux commented on src/Core/3DTiles/C3DTExtensions.js:
```diff
@@ -0,0 +1,88 @@
+            if (this.isExtensionRegistered(extName)) {
+                const extInfo = {};
+                extInfo[extName] = batchTableExtensions[extName].getInfoById(batchID);
+                Object.assign(result, extInfo);
```
You could directly write `result[extName] = batchTableExtensions[extName].getInfoById(batchID);` ?

jailln replied:
Yes, indeed. Done

gchoqueux commented on src/Core/3DTiles/C3DTExtensions.js:
```diff
@@ -0,0 +1,88 @@
+        for (const extName in batchTableExtensions) {
+            // if it is registered it has been parsed and it has a
+            // getInfoByID method
+            if (this.isExtensionRegistered(extName)) {
```
This test `this.isExtensionRegistered(extName)`  isn't necessary?
The classes in `batchTableExtensions ` are instanced only if  extension is already registered?

jailln replied:
Removed.

gchoqueux commented on src/Layer/C3DTilesLayer.js:
```diff
@@ -253,6 +102,63 @@ class C3DTilesLayer extends GeometryLayer {
+        // the  batch table. More details here on itowns internal
+        // organization of 3DTiles:
+        // https://github.com/MEPP-team/RICT/blob/master/Doc/iTowns/Doc.md#itowns-internal-organisation-of-3d-tiles-data
+        const findBatchTable = (object) => {
```
Could you move `findBatchTable` out of `getInfoFromIntersectObject`?

jailln replied:
Done.

gchoqueux commented on src/Core/3DTiles/C3DTBatchTableHierarchyExtension.js:
```diff
@@ -69,46 +73,36 @@ class BatchTableHierarchyExtension extends $3dTilesAbstractExtension {
+            const property = properties[i];
+            instanceProperties[property] =
+                this.classes[instanceClassId].instances[property][instanceIdx];
+        }
```
Could you replace by :
```js
const instances = this.classes[instanceClassId].instances;
for (const key in instances) {
    if (Object.prototype.hasOwnProperty.call(instances, key)) {
        instanceProperties[key] = instances[key][instanceIdx];
    }
}    
```

jailln replied:
Done.

gchoqueux commented on src/Core/3DTiles/C3DTBatchTableHierarchyExtension.js:
```diff
@@ -69,46 +73,36 @@ class BatchTableHierarchyExtension extends $3dTilesAbstractExtension {
+            for (let i = 0; i < this.inverseHierarchy[featureId].length; i++) {
+                const parentID = this.inverseHierarchy[featureId][i];
+                Object.assign(pickingInfo, this.getInfoById(parentID));
+            }
```
could you replace by : 
```js
for (const parentID of  this.inverseHierarchy[featureId]) {
  Object.assign(pickingInfo, this.getInfoById(parentID));
}
```

jailln replied:
Done.

gchoqueux commented on src/Core/3DTiles/C3DTExtensions.js:
```diff
@@ -0,0 +1,88 @@
+        return parsedExtensions;
+    }
+
+    getFeaturesInfoById(batchTableExtensions, batchID) {
```
I'm not sure that the responsibility for acquiring information should be placed in `3DTExtensions`.
It should only handle the registration of extensions and the provisioning of them.

jailln replied:
Moved to `getInfoFromIntersectObject()` of `C3DTilesLayer`.

gchoqueux commented on src/Core/3DTiles/C3DTileset.js:
```diff
@@ -0,0 +1,113 @@
+ */
+class C3DTileset {
+    constructor(json, baseURL, registeredExtensions) {
+        this.type = C3DTilesTypes.tileset;
```
It's more simple to write `'tileset'` directly?

jailln replied:
The idea was to use `C3DTilesTypes` when registering an extension to avoid having to write the same string at different places and relating on the fact they are written in the same way to correctly parse extensions; except that I forgot to use it in `3dtiles_batch_table` example... I modified that.

gchoqueux replied:
`C3DTilesTypes.tileset` it also needs to be written correctly each time and it takes longer to write than `tileset`.

jailln replied:
Well relying on hard-coded strings instead of an enum is generally a bad idea in my opinion... First, `C3DTilesTypes.tileset` being a variable, you are less likely to make a mistake when writting it (e.g. thanks to you IDE). Then, maintenance is easier in case a value must be changed (you only change it in the enum). Finally, you can also expose to the user what values are available by exporting this enum with associated documentation.

On another note, I can write C3DTilesTypes this way: (which is a bit simpler)

````
const C3DTilesTypes = {
    tileset: {},
    batchtable: {},
    boundingVolume: {},
};
````
Anyway, that is your call, let me know.

gchoqueux replied:
You're right. Just keep it like that.

gchoqueux commented on src/Core/3DTiles/C3DTileset.js:
```diff
@@ -0,0 +1,113 @@
+
+        if (json.extensions) {
+            this.extensions =
+                registeredExtensions.parseExtensions(json.extensions, this);
```
is `this.type` to pass?

jailln replied:
Yes, thanks for spotting this; updated.

gchoqueux reviewed (commented):
Yes, that's good, the management of extensions is simpler.

gchoqueux commented on src/Core/3DTiles/C3DTBoundingVolume.js:
```diff
@@ -0,0 +1,128 @@
+        // box[3], box[4], box[5] = x axis direction and half-length
+        // box[6], box[7], box[8] = y axis direction and half-length
+        // box[9], box[10], box[11] = z axis direction and half-length
+        const center = new THREE.Vector3(box[0], box[1], box[2]);
```
Could you move out this class the `center` instance (to avoid Garbage collector).

jailln replied:
Ok, done.

gchoqueux commented on src/Core/3DTiles/C3DTBoundingVolume.js:
```diff
@@ -0,0 +1,128 @@
+        const b = center.z - box[11];
+        const t = center.z + box[11];
+
+        this.box = new THREE.Box3(new THREE.Vector3(w, s, b),
```
Could you move out this class the `Vector3` instance (to avoid Garbage collector).

You can use some `Vector3` instances in all methods.

jailln replied:
Ok, done.

gchoqueux commented on src/Core/3DTiles/C3DTBoundingVolume.js:
```diff
@@ -0,0 +1,128 @@
+    boundingVolumeCulling(camera, tileMatrixWorld) {
+        if (this.region &&
+            !camera.isBox3Visible(this.region.box3D,
+                tileMatrixWorld.clone()
```
Could you instance (before the class) a `Matrix4` and use  `Matrix4.multiplyMatrices` to avoid a clone (and Garbage collector)
```js
matrix.multiplyMatrices(tileMatrixWorld,this.region.matrix);
```

jailln replied:
Ok, done.

gchoqueux commented on src/Core/3DTiles/C3DTExtensions.js:
```diff
@@ -0,0 +1,78 @@
+                if (extManager) {
+                    parsedExtensions[extName] = new extManager(extensionsJSON[extName]);
+                } else {
+                    console.error(`${'No extension manager registered for' +
```
It's not possible to have the error because you check with `isExtensionRegistered`?

jailln replied:
Actually `isExtensionRegistered` only checks for the extension name when it is encountered in the tileset while this test also checks that the extension manager is registered for the `type`. I moved this test into `getExtension`.

gchoqueux commented on src/Layer/C3DTilesLayer.js:
```diff
@@ -253,6 +102,68 @@ class C3DTilesLayer extends GeometryLayer {
+        if (object.parent) {
+            return this.findBatchTable(object.parent);
+        }
+        return undefined;
```
`return undefined;` isn't necessary.

jailln replied:
Ok, removed.

gchoqueux commented on src/Layer/C3DTilesLayer.js:
```diff
@@ -253,6 +102,68 @@ class C3DTilesLayer extends GeometryLayer {
+            // position, normal, batch id)
+            const interAttributes = intersects[i].object.geometry.attributes;
+            if (interAttributes) {
+                const batchIDs = interAttributes._BATCHID || interAttributes.batchId;
```
there is 2 attributes names (`_BATCHID` and `batchId`)  for batch id in the specification ?

jailln replied:
To my knowledge there is only `_BATCHID` but you added `batchId` in [this commit](https://github.com/iTowns/itowns/commit/7b067c7deb21614b731c90e8486ac2fffbcb5e82#diff-276f2190b0230ea4d24aeca3427e133eR25), so I thought there was a good reason to that ? If no, I will remove it since I tried it without `batchId` and it works.

gchoqueux replied:
ok, I'll look into it. We'll leave it like this for now.

gchoqueux reviewed (commented):
I've added a few more comments on the form

gchoqueux commented:
All right, thank you for your efforts.
We'll now test this PR with different use cases.

@gmaillet , @Mkonini could try this PR with with yours 3d tiles applications?

gchoqueux commented on src/Layer/C3DTilesLayer.js:
```diff
@@ -132,7 +131,7 @@ class C3DTilesLayer extends GeometryLayer {
             const interAttributes = intersects[i].object.geometry.attributes;
             if (interAttributes) {
-                const batchIDs = interAttributes._BATCHID || interAttributes.batchId;
+                const batchIDs = interAttributes._BATCHID;
```
could you roll back, i don't remember why i added the `.batchId`

jailln replied:
Done.

zarov commented on src/Core/3DTiles/C3DTileset.js:
```diff
@@ -0,0 +1,113 @@
+ *  [tileset specification](https://github.com/CesiumGS/3d-tiles/blob/master/specification/schema/tileset.schema.json#L11)
+ *  @property {number} geometricError - see [tileset
+ *  specification](https://github.com/CesiumGS/3d-tiles/blob/master/specification/schema/tileset.schema.json#L18)
+ *  @propertu {array} extensionsUsed - see [tileset specification](https://github.com/CesiumGS/3d-tiles/blob/master/specification/schema/tileset.schema.json#L27)
```
Typo `@property`, and could you use `string[]` instead ?

jailln replied:
Done, thanks for spotting that!

jailln commented:
I added some documentation, squashed the commits and rebased on master.

The documentation of `C3DTilesTypes` doesn't get generated, co you have any ideas why ? @zarov maybe ?

gchoqueux commented:
We revealed a issue (may be on screen space error).  We had try the same data  with this PR, the buildings wasn't displayed. Have you try with yours datas?

zarov commented:
> The documentation of `C3DTilesTypes` doesn't get generated, co you have any ideas why ? @zarov maybe ?

You can do something similar to [`CACHE_POLICIES` here](https://github.com/iTowns/itowns/blob/master/src/Core/Scheduler/Cache.js#L3-L19).

gchoqueux commented:
I find a problem, the bounding boxes are wrong:

#### Master
![bbox master](https://user-images.githubusercontent.com/11291849/85026737-55640980-b179-11ea-8eb0-e76adb81b0c2.png)

#### PR
![bbox PR](https://user-images.githubusercontent.com/11291849/85026757-5bf28100-b179-11ea-9335-6a7152e43ec2.png)

jailln commented:
> You can do something similar to [`CACHE_POLICIES` here](https://github.com/iTowns/itowns/blob/master/src/Core/Scheduler/Cache.js#L3-L19).

It doesn't seem to work (see my last commit). I also tried to change the default export to a named export but it work either.

@gchoqueux, thanks for finding that! I will look into it.

zarov commented:
> It doesn't seem to work (see my last commit). I also tried to change the default export to a named export but it work either.

It does not indeed, in fact it should use `@enum` instead of `@typedef`, my bad. However, enums are not correctly generated in the current documentation. What you can do is makes this change, but remove it from `docs/config.json` for now. I'll try to find some times to add it in the documentation, sorry about that.

jailln commented:
@gchoqueux Removing global variables from C3DTBoundingVolume fixes this bug. I guess we should avoid them...

@zarov, ok thanks!

gchoqueux commented:
It's not necessary to remove all global variables:
* in `initBoundingRegion`:
  just keep `const regionBox = new OBB();` because it needs to instance a new `OBB` for each `this.region`.
* in `initBoundingBox` replace all code by:
```js
center.set(box[0], box[1], box[2]);
size.set(box[3], box[7], box[11]).multiplyScalar(2);
this.box = new THREE.Box3();
this.box.setFromCenterAndSize(center, size);
```

jailln commented:
> It's not necessary to remove all global variables:
> 
>     * in `initBoundingRegion`:
>       just keep `const regionBox = new OBB();` because it needs to instance a new `OBB` for each `this.region`.
> 
>     * in `initBoundingBox` replace all code by:
> 
> 
> ```js
> center.set(box[0], box[1], box[2]);
> size.set(box[3], box[7], box[11]).multiplyScalar(2);
> this.box = new THREE.Box3();
> this.box.setFromCenterAndSize(center, size);
> ```

Ok, done.

gchoqueux commented:
Ok, thanks
Could you resolve resolve the conflict?
I re-check with other datas and it'll be over soon.

jailln commented:
> Ok, thanks
> Could you resolve resolve the conflict?
> I re-check with other datas and it'll be over soon.

Squashed and rebased.

gchoqueux reviewed (approved):
Thank you for your work!

zarov commented:
Well done @jailln :partying_face:

jailln commented:
Nice, thanks! :partying_face:
